text/0000-render-element-modifiers.md

```diff
@@ -0,0 +1,587 @@
+
+```js
+export default Component.extend({
+  tagName: '',
+
+  didInsertList(element) {
+    // ...
+  },
+
+  willDestroyList(element) {
+    // ...
+  }
+});
+```
+```hbs
+<ul
+  {{did-render (action this.didInsertList)}}
+  {{will-destroy (action this.willDestroyList)}}
+>
+  ...
+</ul>
+
+<div>
+  ...
+</div>
+```
+
+### Reusable Helpers
+
+Currently, the best ways to share element setup code are either via mixins,
+which are somewhat opaque and can encourage problematic patterns, or standard JS
+functions, which generally require some amount of boilerplate.
+
+Developers will be able to define element modifiers in the future with modifier
+managers provided by addons. However, the proposed modifier APIs are fairly
+verbose (with good reason) and not stabilized.
+
+However, `{{did-render}}` and `{{will-destroy}}` can receive _any_ function as
+their first parameter, allowing users to share and reuse common element setup
+code with helpers. For instance, a simple `scrollTo` helper could be created to
+set the scroll position of an element:
+
+```js
+// helpers/scroll-to.js
+export default function scrollTo() {
+  (element, scrollPosition) => element.scrollTop = scrollPosition;
+}
+```
+```hbs
+<div {{did-render (scroll-to) @scrollPosition}} class="scroll-container">
+  ...
+</div>
+```
+
+## Detailed design
+
+This RFC proposes adding two element modifiers, `{{did-render}}` and
+`{{will-destroy}}`. Note that element modifiers do _not_ run in SSR mode - this
+code is only run on clients.
+
+> Note: The timing semantics in the following section were mostly defined in the
+> [element modifier manager RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+> and are repeated here for clarity and convenience.
+
+### `{{did-render}}`
+
+This modifier is activated:
+
+1. When The element is inserted in the DOM
+2. Whenever any of the arguments passed to it update, including the function
+   passed as the first argument.
+
+It has the following timing semantics when activated:
+
+* **Always**
+  * called after DOM insertion
+  * called _after_ any child element's `{{did-render}}` modifiers
+  * called _after_ the enclosing component's `willRender` hook
+  * called _before_ the enclosing component's `didRender` hook
+  * called in definition order in the template
+  * called after the arguments to the modifier have changed
+* **May or May Not**
+  * be called in the same tick as DOM insertion
+  * have the sibling nodes fully initialized in DOM
+* **Never**
+  * called if the arguments to the modifier are constants
+
+Note that these statements do not refer to when the modifier is _activated_,
+only to when it will be run relative to other hooks and modifiers _should it be
+activated_. The modifier is only activated on insertion and arg changes.
+
+`{{did-render}}` receives a function with the following signature as the first
+positional parameter:
+
+```ts
+type DidRenderHandler = (element: Element, ...args): void;
```

Comment:
Should this take `params` and `hash` like helpers? Otherwise, any named arguments that are passed will be "lost".

Reply:
Updated to account for that, thanks for the suggestion @chancancode!